### ISSUE TITLE
How to get Last changes from datastore on basis of sequence number?

### ISSUE THREAD

**Issue by SakshiJainAirfi** [cloudant/CDTDatastore#457]
As in Android sdk, gives API called 
```
datastore.changes(SEQUENCE NUMBER)
datastore.getLastSequence()
```

Does CDTDatastore sameAPI's for iOS?

**Comment by ricellis** [comment]
[`CDTFetchChanges`](https://github.com/cloudant/CDTDatastore/blob/master/CDTDatastore/CDTFetchChanges.h) is what you want. The interface is pretty well documented, but basically you can set blocks to get called for:
* [`documentChanged`](https://github.com/cloudant/CDTDatastore/blob/master/CDTDatastore/CDTFetchChanges.h#L112)
* [`documentWithIDWasDeletedBlock`](https://github.com/cloudant/CDTDatastore/blob/master/CDTDatastore/CDTFetchChanges.h#L132)
* [`fetchRecordChangesCompletionBlock`](https://github.com/cloudant/CDTDatastore/blob/master/CDTDatastore/CDTFetchChanges.h#L161) - this one gets called with the last sequence id

**Comment by SakshiJainAirfi** [comment]
Thanks for the reply..!!

But I want to know 2 things more:- 

1. From where should we fetch **last sequence id**?

2. Can I get **code example** for all above methods?

**Comment by ricellis** [comment]
1. You supply a `fetchRecordChangeCompletionBlock` that gets called with the last sequence ID, it's called `newSequenceValue` there (as per 2, [the tests do this in their setup method](https://github.com/cloudant/CDTDatastore/blob/a21d253bf0f2e50af68ea1c41578778210d17120/CDTDatastoreTests/CDTFetchChangesTests.m#L52-L55)).
2. I don't think we have a specific example, but you can see them in use in the [tests](https://github.com/cloudant/CDTDatastore/blob/master/CDTDatastoreTests/CDTFetchChangesTests.m).

**Comment by SakshiJainAirfi** [comment]
I was in between of implementation of **documentChangedBlock** in swift and things I noticed - 

- Any change in variable inside made inside this block can't get outside the block
- there is no way to get when this block has finished execution
- this block starts with delay.
- there is no way to know whether the block will start executing or not?

**Comment by ricellis** [comment]
>Any change in variable inside made inside this block can't get outside the block

Variables need special scope for blocks, see [blocks documentation](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Blocks/Articles/00_Introduction.html#//apple_ref/doc/uid/TP40007502-CH1-SW1). I'm not sure what the equivalent to `__block` is in Swift.

>there is no way to get when this block has finished execution

The `documentChangedBlock` is called once per (non-deletion) document change. After the operation starts (either by calling `start` directly or when it is picked off the `NSOperationQueue`  where it was submitted) then the block will be invoked for each document change in the local datastore changes feed until the results limit (default 500) is reached, there are no further changes, or the operation is cancelled.

If the block hasn't been called before `fetchRecordChangesCompletionBlock` then there are no non-deleted document changes less than the results limit. You can take the last seq and call for the next set of changes.

>this block starts with delay.

The `documentChangedBlock` may start with a delay because there are other changes (e.g. deleted documents) earlier in the changes feed that are processed first or also because of waiting for a thread for execution from the queue.

>there is no way to know whether the block will start executing or not? If not then, we need to execute some other functionality!!

The block will start executing as long as you set the result limit high enough or repeatedly fetch changes from new last seq values until there are no further changes left and as long as there are changes of the appropriate type in the local datastore to fetch.

If what you're actually trying to do is operate as a listener on a datastore then you should use the [event style `NSNotificationCenter`](https://github.com/cloudant/CDTDatastore/blob/master/doc/events.md) instead.

**Comment by SakshiJainAirfi** [comment]
> Variables need special scope for blocks, see [blocks documentation](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Blocks/Articles/00_Introduction.html#//apple_ref/doc/uid/TP40007502-CH1-SW1). I'm not sure what the equivalent to `__block` is in Swift

I agree variable needs special scope in objectiveC while using in blocks.But In Swift, there is no special way required to save data inside from a block.**What I need is to get unsynced document count?**
Which currently seems to be impossible from this `documentChangedBlock` because it is a continous block and continuously throwing notification and updating on UI doesn't seems to be efficient solution.

**Comment by ricellis** [comment]
There is really no such thing as an "unsynced document count" - you can _approximate_ the number of local changes that haven't been replicated I suppose by calling for the local datastore changes since the last push replication, assuming no pull replication also modifies the database (because those changes would show up locally, but not need syncing to a remote). Something like

```
// 1. At the end of a replication get the current sequence value with a fetchRecordChangesCompletionBlock
CDTFetchChanges *fetchChanges =
    [[CDTFetchChanges alloc] initWithDatastore:self.datastore startSequenceValue:nil];

__block NSString *blockStartSequenceValue = nil;
fetchChanges.fetchRecordChangesCompletionBlock =
    ^(NSString *newSequenceValue, NSString *startSequenceValue, NSError *fetchError) {
      blockStartSequenceValue = newSequenceValue;
    };

// run synchronously
[fetchChanges start];

self.startSequenceValue = blockStartSequenceValue;

// 2. Count the number of changes since that sequence number
CDTFetchChanges *fetchChanges =
    [[CDTFetchChanges alloc] initWithDatastore:self.datastore
                            startSequenceValue:self.startSequenceValue];

__block NSUInteger blockChangedDocCount = 0;
fetchChanges.documentChangedBlock = ^(CDTDocumentRevision *revision) {
  blockChangedDocCount++;
};

__block NSUInteger blockDeletedDocCount = 0;
fetchChanges.documentWithIDWasDeletedBlock = ^(NSString *docId) {
  blockDeletedDocCount++;
};

fetchChanges.fetchRecordChangesCompletionBlock =
    ^(NSString *newSequenceValue, NSString *startSequenceValue, NSError *fetchError) {
      blockStartSequenceValue = newSequenceValue;
    };

// run synchonously
[fetchChanges start];
```

I think you'll want to put step 2 in a loop checking for `fetchChanges.moreComing` to page through each set of 500 changes (or fewer if you customize the result limit) unless you are sure that you'll always have fewer changes than that when you run this. When `moreComing` is NO/false you've reached the equivalent of "now".

**Comment by glynnbird** [comment]
Further commentary provided by the ticket creator via email:

> The issues are like in these API's DocumentChangedBlock and fetchRecordChangesCompletionBlock are continuous blocks, which means if we have multiple DB's and I need unsynced document count for all DB's (in a loop) there I wont get to know for which DB's count is coming because these block API's 
work like typical ObjectiveC blocks, where latency and retaining  value inside and outside a block becomes issue.

**Comment by SakshiJainAirfi** [comment]
Via Email:

> I think the code (and the guidance) in the ticket shows how the number of changes can be counted for a single database. If you need the answer for multiple databases, then isn't the answer to create a new CDTFetchChanges object for each database in series?
> As I say, this isn't an API that is provided by the service directly. If you need to maintain a count of documents you have on a device that haven't yet been synced, then would it be possible to maintain a counter in your app for this purpose?

**Comment by SakshiJainAirfi** [comment]
Via Email:

> API is already for their DocumentChangedBlock, just the issue is how it is working!!
> Instead of block functionality, a method can return us Array of RevisionsId ,which will solve the issue.
> In Android, datastore.changes API works in same way.

**Comment by ricellis** [comment]
I don't understand the problem - if you want to synchronously get an array of all the changes from `someSequence` until "now" just collect them from the block:
```obj-c
    CDTFetchChanges *fetchChanges =
        [[CDTFetchChanges alloc] initWithDatastore:self.datastore
                                startSequenceValue:someSequence];

    __block NSMutableArray *revs = [[NSMutableArray alloc] init];
    fetchChanges.documentChangedBlock = ^(CDTDocumentRevision *revision) {
        [revs addObject: revision];
    };

    [fetchChanges start];

    for (CDTDocumentRevision *doc in revs) {
        // Do something with each rev
        printf("%s %s \n", [doc.docId UTF8String], [doc.revId UTF8String]);
    }

```
